controllers/raycluster_controller.go

```diff
@@ -0,0 +1,325 @@
+	_, err = r.kubeClient.CoreV1().Secrets(cluster.Namespace).Apply(ctx, desiredOAuthSecret(&cluster, r), metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+	if err != nil {
+		logger.Error(err, "Failed to create OAuth Secret")
+	}
+
+	_, err = r.kubeClient.CoreV1().Services(cluster.Namespace).Apply(ctx, desiredOAuthService(&cluster), metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+	if err != nil {
+		logger.Error(err, "Failed to update OAuth Service")
+	}
+
+	_, err = r.kubeClient.CoreV1().ServiceAccounts(cluster.Namespace).Apply(ctx, desiredServiceAccount(&cluster), metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+	if err != nil {
+		logger.Error(err, "Failed to update OAuth ServiceAccount")
+	}
+
+	_, err = r.kubeClient.RbacV1().ClusterRoleBindings().Apply(ctx, desiredOAuthClusterRoleBinding(&cluster), metav1.ApplyOptions{FieldManager: controllerName, Force: true})
+	if err != nil {
+		logger.Error(err, "Failed to update OAuth ClusterRoleBinding")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func crbNameFromCluster(cluster *rayv1.RayCluster) string {
+	return cluster.Name + "-" + cluster.Namespace + "-auth" // NOTE: potential naming conflicts ie {name: foo, ns: bar-baz} and {name: foo-bar, ns: baz}
+}
+
+func desiredOAuthClusterRoleBinding(cluster *rayv1.RayCluster) *rbacapply.ClusterRoleBindingApplyConfiguration {
+	return rbacapply.ClusterRoleBinding(
+		crbNameFromCluster(cluster)).
+		WithLabels(map[string]string{"ray.io/cluster-name": cluster.Name}).
+		WithSubjects(
+			rbacapply.Subject().
+				WithKind("ServiceAccount").
+				WithName(cluster.Name).
+				WithNamespace(cluster.Namespace),
+		).
+		WithRoleRef(
+			rbacapply.RoleRef().
+				WithAPIGroup("rbac.authorization.k8s.io").
+				WithKind("ClusterRole").
+				WithName("system:auth-delegator"),
+		).
+		WithOwnerReferences(
+			v1.OwnerReference().WithUID(cluster.UID).WithName(cluster.Name).WithKind(cluster.Kind).WithAPIVersion(cluster.APIVersion),
+		)
+}
+
+func desiredServiceAccount(cluster *rayv1.RayCluster) *coreapply.ServiceAccountApplyConfiguration {
+	return coreapply.ServiceAccount(cluster.Name, cluster.Namespace).
+		WithLabels(map[string]string{"ray.io/cluster-name": cluster.Name}).
+		WithAnnotations(map[string]string{
+			"serviceaccounts.openshift.io/oauth-redirectreference.first": "" +
+				`{"kind":"OAuthRedirectReference","apiVersion":"v1",` +
+				`"reference":{"kind":"Route","name":"` + cluster.Name + `"}}`,
+		}).
+		WithOwnerReferences(
+			v1.OwnerReference().WithUID(cluster.UID).WithName(cluster.Name).WithKind(cluster.Kind).WithAPIVersion(cluster.APIVersion),
+		)
+}
+
+func desiredClusterRoute(cluster *rayv1.RayCluster) *routeapply.RouteApplyConfiguration {
+	return routeapply.Route(cluster.Name, cluster.Namespace).
+		WithLabels(map[string]string{"ray.io/cluster-name": cluster.Name}).
+		WithSpec(routeapply.RouteSpec().
+			WithTo(routeapply.RouteTargetReference().WithKind("Service").WithName(cluster.Name + "tls")).
```

Comment:
Where is that service created? I'd advocate to have it created by this controller as well.

Reply:
https://github.com/project-codeflare/codeflare-operator/pull/453/files#diff-9514176087ba6764592f4797d22a2ef7e9c4b0204ecfc94efa00d371b66f6bb7R271

It's already created by this controller

Reply:
Ah right, got confused somehow :).